File: frontend/src/constants/chordSuffixes.js
/**
 * Canonical chord-suffix vocabulary — the single source of truth for the app.
 *
 * Every other suffix table (the ChordPro detection regex, the chords-db lookup
 * map, and the all-guitar-chords URL map) is DERIVED from this list, so they
 * can never drift apart. The Python backend mirrors this file in
 * `backend/services/chord_suffixes.py`; keep the two in sync.
 *
 * Each entry:
 *   canonical : canonical label used when displaying / simplifying a chord
 *   aliases   : every sheet spelling that resolves to this suffix
 *               (the empty string '' means a bare major chord, e.g. "C")
 *   db        : @tombatossals/chords-db `suffix`, or null if absent there
 *   agc       : all-guitar-chords.com URL segment, or null if unavailable
 *   intervals : semitones from the root (used by the movable-shape generator)
 *   parent    : simpler suffix to fall back to when no exact voicing exists.
 *               Walking `parent` repeatedly always terminates at a core family
 *               ('major' / 'minor') that has a guaranteed voicing. A non-null
 *               parent means a fallback diagram is a *simplification* and must
 *               be flagged with a trailing '*' in the UI.
 */
export const CHORD_SUFFIXES = [
  // --- Triads -------------------------------------------------------------
  { canonical: 'major', aliases: ['', 'maj', 'M', 'major'], db: 'major', agc: 'major', intervals: [0, 4, 7], parent: null },
  { canonical: 'minor', aliases: ['m', 'min', 'minor'], db: 'minor', agc: 'minor', intervals: [0, 3, 7], parent: null },
  { canonical: 'dim', aliases: ['dim'], db: 'dim', agc: 'dim', intervals: [0, 3, 6], parent: 'minor' },
  { canonical: 'aug', aliases: ['aug'], db: 'aug', agc: 'aug', intervals: [0, 4, 8], parent: 'major' },
  { canonical: 'sus2', aliases: ['sus2'], db: 'sus2', agc: 'sus2', intervals: [0, 2, 7], parent: 'major' },
  { canonical: 'sus4', aliases: ['sus4', 'sus'], db: 'sus4', agc: 'sus4', intervals: [0, 5, 7], parent: 'major' },

  // --- Sixths -------------------------------------------------------------
  { canonical: '6', aliases: ['6'], db: '6', agc: '6', intervals: [0, 4, 7, 9], parent: 'major' },
  { canonical: 'm6', aliases: ['m6', 'min6'], db: 'm6', agc: 'm6', intervals: [0, 3, 7, 9], parent: 'minor' },

  // --- Sevenths -----------------------------------------------------------
  { canonical: '7', aliases: ['7', 'dom7'], db: '7', agc: '7', intervals: [0, 4, 7, 10], parent: 'major' },
  { canonical: 'maj7', aliases: ['maj7', 'M7', 'major7'], db: 'maj7', agc: 'maj7', intervals: [0, 4, 7, 11], parent: 'major' },
  { canonical: 'm7', aliases: ['m7', 'min7', 'minor7'], db: 'm7', agc: 'm7', intervals: [0, 3, 7, 10], parent: 'minor' },
  { canonical: 'm7b5', aliases: ['m7b5', 'min7b5', 'minor7b5'], db: 'm7b5', agc: 'm7b5', intervals: [0, 3, 6, 10], parent: 'dim' },
  { canonical: 'mmaj7', aliases: ['mmaj7', 'mM7', 'minmaj7'], db: 'mmaj7', agc: null, intervals: [0, 3, 7, 11], parent: 'm7' },
  { canonical: 'dim7', aliases: ['dim7'], db: 'dim7', agc: 'dim7', intervals: [0, 3, 6, 9], parent: 'dim' },

  // --- Altered dominants --------------------------------------------------
  { canonical: '7b5', aliases: ['7b5'], db: '7b5', agc: '7b5', intervals: [0, 4, 6, 10], parent: '7' },
  { canonical: '7#5', aliases: ['7#5', 'aug7', '7+5'], db: 'aug7', agc: null, intervals: [0, 4, 8, 10], parent: '7' },
  { canonical: '7b9', aliases: ['7b9'], db: '7b9', agc: '7b9', intervals: [0, 4, 7, 10, 13], parent: '7' },
  { canonical: '7#9', aliases: ['7#9'], db: '7#9', agc: null, intervals: [0, 4, 7, 10, 15], parent: '7' },
  { canonical: '7sus4', aliases: ['7sus4'], db: '7sus4', agc: '7sus4', intervals: [0, 5, 7, 10], parent: 'sus4' },
  { canonical: 'm7#5', aliases: ['m7#5'], db: null, agc: null, intervals: [0, 3, 8, 10], parent: 'm7' },
  { canonical: 'maj7b5', aliases: ['maj7b5'], db: 'maj7b5', agc: null, intervals: [0, 4, 6, 11], parent: 'maj7' },
  { canonical: 'maj7#5', aliases: ['maj7#5'], db: 'maj7#5', agc: null, intervals: [0, 4, 8, 11], parent: 'maj7' },

  // --- Ninths -------------------------------------------------------------
  { canonical: '9', aliases: ['9'], db: '9', agc: '9', intervals: [0, 4, 7, 10, 14], parent: '7' },
  { canonical: 'm9', aliases: ['m9', 'min9'], db: 'm9', agc: 'm9', intervals: [0, 3, 7, 10, 14], parent: 'm7' },
  { canonical: 'maj9', aliases: ['maj9', 'M9'], db: 'maj9', agc: null, intervals: [0, 4, 7, 11, 14], parent: 'maj7' },

  // --- Elevenths ----------------------------------------------------------
  { canonical: '11', aliases: ['11'], db: '11', agc: '11', intervals: [0, 7, 10, 14, 17], parent: '9' },
  { canonical: 'm11', aliases: ['m11', 'min11'], db: 'm11', agc: 'm11', intervals: [0, 3, 7, 10, 14, 17], parent: 'm9' },
  { canonical: 'maj11', aliases: ['maj11', 'M11'], db: 'maj11', agc: null, intervals: [0, 4, 7, 11, 14, 17], parent: 'maj9' },

  // --- Thirteenths --------------------------------------------------------
  { canonical: '13', aliases: ['13'], db: '13', agc: '13', intervals: [0, 4, 7, 10, 14, 21], parent: '9' },
  { canonical: 'm13', aliases: ['m13', 'min13'], db: null, agc: null, intervals: [0, 3, 7, 10, 14, 21], parent: 'm9' },
  { canonical: 'maj13', aliases: ['maj13', 'M13'], db: 'maj13', agc: null, intervals: [0, 4, 7, 11, 14, 21], parent: 'maj9' },

  // --- Added tones --------------------------------------------------------
  { canonical: 'add9', aliases: ['add9'], db: 'add9', agc: 'add9', intervals: [0, 4, 7, 14], parent: 'major' },
  { canonical: 'add2', aliases: ['add2'], db: null, agc: null, intervals: [0, 2, 4, 7], parent: 'major' },
  { canonical: 'add4', aliases: ['add4'], db: null, agc: null, intervals: [0, 4, 5, 7], parent: 'major' },
  { canonical: 'add11', aliases: ['add11'], db: 'add11', agc: null, intervals: [0, 4, 7, 17], parent: 'major' },
  // Bare "add" (e.g. "Cadd") is malformed but appears in real sheets — treat as major.
  { canonical: 'add', aliases: ['add'], db: 'major', agc: 'major', intervals: [0, 4, 7], parent: 'major' },
];

/** canonical suffix -> entry (fast lookup for the parent chain etc.). */
export const SUFFIX_BY_CANONICAL = new Map(CHORD_SUFFIXES.map((entry) => [entry.canonical, entry]));

function buildAliasMap(field) {
  const map = {};
  for (const entry of CHORD_SUFFIXES) {
    for (const alias of entry.aliases) {
      if (entry[field] != null && map[alias] == null) {
        map[alias] = entry[field];
      }
    }
  }
  return map;
}

/** Sheet suffix -> chords-db `suffix`. */
export const SUFFIX_TO_DB = buildAliasMap('db');

/** Sheet suffix -> all-guitar-chords.com URL segment. */
export const SUFFIX_TO_AGC = buildAliasMap('agc');

/** Sheet suffix -> canonical suffix (for simplification + display). */
export const SUFFIX_TO_CANONICAL = (() => {
  const map = {};
  for (const entry of CHORD_SUFFIXES) {
    for (const alias of entry.aliases) {
      if (map[alias] == null) map[alias] = entry.canonical;
    }
  }
  return map;
})();

/**
 * Aliases excluded from chord-token *detection* because they collide with
 * ordinary text (e.g. a bare "M" would flag "AM"/"GM" as chords). They remain
 * valid for diagram resolution if a chord is detected by other means.
 */
export const NON_DETECTABLE_ALIASES = new Set(['M']);

/**
 * All non-empty, detectable aliases, sorted longest-first so the ChordPro
 * tokenizer matches greedily (e.g. `maj7` before `maj`, `m7b5` before `m7`,
 * `min7` before `min`).
 */
export const SUFFIX_ALIASES_BY_LENGTH = CHORD_SUFFIXES.flatMap((entry) => entry.aliases)
  .filter((alias) => alias !== '' && !NON_DETECTABLE_ALIASES.has(alias))
  .sort((a, b) => b.length - a.length);

function escapeRegExp(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Regex source for an optional chord suffix, derived from the canonical list.
 * Used to build the chord-token detection pattern in chordpro.js.
 */
export const CHORD_SUFFIX_PATTERN = `(?:${SUFFIX_ALIASES_BY_LENGTH.map(escapeRegExp).join('|')})?`;

/**
 * Resolve the simplification chain for a canonical suffix: returns the ordered
 * list of canonical suffixes to try, from the exact chord down to its core
 * triad. The first element is the suffix itself.
 */
export function suffixFallbackChain(canonical) {
  const chain = [];
  let current = canonical;
  const seen = new Set();
  while (current && !seen.has(current)) {
    seen.add(current);
    chain.push(current);
    const entry = SUFFIX_BY_CANONICAL.get(current);
    current = entry?.parent ?? null;
  }
  return chain;
}

File: backend/services/chord_suffixes.py
"""Canonical chord-suffix vocabulary (Python mirror).

This mirrors `frontend/src/constants/chordSuffixes.js`. Keep the two in sync;
the styler/coverage tests assert that the derived maps match expectations.

Each entry is (canonical, aliases, db, agc, intervals, parent). See the JS file
for the full field documentation.
"""

from __future__ import annotations

from typing import NamedTuple


class SuffixEntry(NamedTuple):
    canonical: str
    aliases: tuple[str, ...]
    db: str | None
    agc: str | None
    intervals: tuple[int, ...]
    parent: str | None


CHORD_SUFFIXES: tuple[SuffixEntry, ...] = (
    # Triads
    SuffixEntry("major", ("", "maj", "M", "major"), "major", "major", (0, 4, 7), None),
    SuffixEntry("minor", ("m", "min", "minor"), "minor", "minor", (0, 3, 7), None),
    SuffixEntry("dim", ("dim",), "dim", "dim", (0, 3, 6), "minor"),
    SuffixEntry("aug", ("aug",), "aug", "aug", (0, 4, 8), "major"),
    SuffixEntry("sus2", ("sus2",), "sus2", "sus2", (0, 2, 7), "major"),
    SuffixEntry("sus4", ("sus4", "sus"), "sus4", "sus4", (0, 5, 7), "major"),
    # Sixths
    SuffixEntry("6", ("6",), "6", "6", (0, 4, 7, 9), "major"),
    SuffixEntry("m6", ("m6", "min6"), "m6", "m6", (0, 3, 7, 9), "minor"),
    # Sevenths
    SuffixEntry("7", ("7", "dom7"), "7", "7", (0, 4, 7, 10), "major"),
    SuffixEntry("maj7", ("maj7", "M7", "major7"), "maj7", "maj7", (0, 4, 7, 11), "major"),
    SuffixEntry("m7", ("m7", "min7", "minor7"), "m7", "m7", (0, 3, 7, 10), "minor"),
    SuffixEntry("m7b5", ("m7b5", "min7b5", "minor7b5"), "m7b5", "m7b5", (0, 3, 6, 10), "dim"),
    SuffixEntry("mmaj7", ("mmaj7", "mM7", "minmaj7"), "mmaj7", None, (0, 3, 7, 11), "m7"),
    SuffixEntry("dim7", ("dim7",), "dim7", "dim7", (0, 3, 6, 9), "dim"),
    # Altered dominants
    SuffixEntry("7b5", ("7b5",), "7b5", "7b5", (0, 4, 6, 10), "7"),
    SuffixEntry("7#5", ("7#5", "aug7", "7+5"), "aug7", None, (0, 4, 8, 10), "7"),
    SuffixEntry("7b9", ("7b9",), "7b9", "7b9", (0, 4, 7, 10, 13), "7"),
    SuffixEntry("7#9", ("7#9",), "7#9", None, (0, 4, 7, 10, 15), "7"),
    SuffixEntry("7sus4", ("7sus4",), "7sus4", "7sus4", (0, 5, 7, 10), "sus4"),
    SuffixEntry("m7#5", ("m7#5",), None, None, (0, 3, 8, 10), "m7"),
    SuffixEntry("maj7b5", ("maj7b5",), "maj7b5", None, (0, 4, 6, 11), "maj7"),
    SuffixEntry("maj7#5", ("maj7#5",), "maj7#5", None, (0, 4, 8, 11), "maj7"),
    # Ninths
    SuffixEntry("9", ("9",), "9", "9", (0, 4, 7, 10, 14), "7"),
    SuffixEntry("m9", ("m9", "min9"), "m9", "m9", (0, 3, 7, 10, 14), "m7"),
    SuffixEntry("maj9", ("maj9", "M9"), "maj9", None, (0, 4, 7, 11, 14), "maj7"),
    # Elevenths
    SuffixEntry("11", ("11",), "11", "11", (0, 7, 10, 14, 17), "9"),
    SuffixEntry("m11", ("m11", "min11"), "m11", "m11", (0, 3, 7, 10, 14, 17), "m9"),
    SuffixEntry("maj11", ("maj11", "M11"), "maj11", None, (0, 4, 7, 11, 14, 17), "maj9"),
    # Thirteenths
    SuffixEntry("13", ("13",), "13", "13", (0, 4, 7, 10, 14, 21), "9"),
    SuffixEntry("m13", ("m13", "min13"), None, None, (0, 3, 7, 10, 14, 21), "m9"),
    SuffixEntry("maj13", ("maj13", "M13"), "maj13", None, (0, 4, 7, 11, 14, 21), "maj9"),
    # Added tones
    SuffixEntry("add9", ("add9",), "add9", "add9", (0, 4, 7, 14), "major"),
    SuffixEntry("add2", ("add2",), None, None, (0, 2, 4, 7), "major"),
    SuffixEntry("add4", ("add4",), None, None, (0, 4, 5, 7), "major"),
    SuffixEntry("add11", ("add11",), "add11", None, (0, 4, 7, 17), "major"),
    SuffixEntry("add", ("add",), "major", "major", (0, 4, 7), "major"),
)

SUFFIX_BY_CANONICAL: dict[str, SuffixEntry] = {e.canonical: e for e in CHORD_SUFFIXES}


def _build_alias_map(field: str) -> dict[str, str]:
    out: dict[str, str] = {}
    for entry in CHORD_SUFFIXES:
        value = getattr(entry, field)
        if value is None:
            continue
        for alias in entry.aliases:
            out.setdefault(alias, value)
    return out


SUFFIX_TO_DB = _build_alias_map("db")
SUFFIX_TO_AGC = _build_alias_map("agc")


def _build_canonical_map() -> dict[str, str]:
    out: dict[str, str] = {}
    for entry in CHORD_SUFFIXES:
        for alias in entry.aliases:
            out.setdefault(alias, entry.canonical)
    return out


SUFFIX_TO_CANONICAL = _build_canonical_map()


def suffix_fallback_chain(canonical: str) -> list[str]:
    """Ordered simplification chain from `canonical` down to its core triad."""
    chain: list[str] = []
    current: str | None = canonical
    seen: set[str] = set()
    while current and current not in seen:
        seen.add(current)
        chain.append(current)
        entry = SUFFIX_BY_CANONICAL.get(current)
        current = entry.parent if entry else None
    return chain
